File: python/intervels/Non-overlappingIntervals.py
'''
Given an array of intervals intervals where intervals[i] = [starti, endi], return the minimum number of intervals you need to remove to
 make the rest of the intervals non-overlapping.

Note that intervals which only touch at a point are non-overlapping. For example, [1, 2] and [2, 3] are non-overlapping.

Example 1:
Input: intervals = [[1,2],[2,3],[3,4],[1,3]]
Output: 1
Explanation: [1,3] can be removed and the rest of the intervals are non-overlapping.

Example 2:
Input: intervals = [[1,2],[1,2],[1,2]]
Output: 2
Explanation: You need to remove two [1,2] to make the rest of the intervals non-overlapping.

Example 3:
Input: intervals = [[1,2],[2,3]]
Output: 0
Explanation: You don't need to remove any of the intervals since they're already non-overlapping.
'''
from typing import List

class NonOverlappingIntervals:
    def eraseOverlapIntervalsNeet(self, intervals: List[List[int]]) -> int:
        intervals.sort()
        #intervals.sort(key = lambda x : x[1])
        res = 0
        prevEnd = intervals[0][1]
        for start, end in intervals[1:]:
            if start >= prevEnd:
                prevEnd = end
            else:
                res += 1
            prevEnd =  min(end, prevEnd) #end

        return res

    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0
        # Sort by end time to maximize the number of non-overlapping intervals we can keep
        intervals.sort(key=lambda x: x[1])
        count = 0
        prev_end = intervals[0][1]
        for start, end in intervals[1:]:
            # If overlap occurs (start is less than previous end)
            if start < prev_end:
                count += 1
                # We effectively "remove" the current interval by not updating prev_end
                # because the current one ends later (or same time) and is less optimal
            else:
                # No overlap, update end time to current interval's end
                prev_end = end
        return count

if __name__ == "__main__":
    solution = NonOverlappingIntervals()
    
    # Test Case 1
    intervals1 = [[1,2],[2,3],[3,4],[1,3]]
    print(f"Input: {intervals1}")
    print(f"Output: {solution.eraseOverlapIntervalsNeet(intervals1)}") # Expected: 1
    print("-" * 20)

    # Test Case 2
    intervals2 = [[1,2],[1,2],[1,2]]
    print(f"Input: {intervals2}")
    print(f"Output: {solution.eraseOverlapIntervalsNeet(intervals2)}") # Expected: 2
    print("-" * 20)

    # Test Case 3
    intervals3 = [[1,2],[2,3]]
    print(f"Input: {intervals3}")
    print(f"Output: {solution.eraseOverlapIntervalsNeet(intervals3)}") # Expected: 0

File: src/main/java/com/naresh/a_dsalgo/Intervals/problems/NonOverlappingIntervals.java
package com.naresh.a_dsalgo.Intervals.problems;

import java.util.Arrays;
import java.util.Comparator;

/**
 * Problem: Non-overlapping Intervals (LeetCode 435)
 * Description: Given an array of intervals `intervals` where `intervals[i] = [starti, endi]`,
 * return the minimum number of intervals you need to remove to make the rest of the intervals non-overlapping.
 */
public class NonOverlappingIntervals {

    /**
     * Algorithm: Greedy approach. Sort intervals by their end times. Iterate through the sorted intervals,
     * keeping track of the end time of the last non-overlapping interval found (`prevEnd`). If a current
     * interval overlaps with `prevEnd`, it must be removed (increment `removedCount`). Otherwise, update `prevEnd`.
     *
     * @param intervals A 2D array of intervals `[start, end]`.
     * @return The minimum number of intervals to remove.
     */
    public int eraseOverlapIntervals(int[][] intervals) {
        // Pattern: Greedy (Sorting) | Time: O(N log N), Space: O(1)
        if (intervals == null || intervals.length <= 1) return 0;

        // Sort intervals by their end times. This is the greedy choice:
        // By picking the interval that finishes earliest, we leave maximum room for subsequent intervals.
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[1]));

        var removedCount = 0;
        var prevEnd = intervals[0][1]; // End time of the first non-overlapping interval chosen

        // Iterate from the second interval
        for (var i = 1; i < intervals.length; i++) {
            var currentStart = intervals[i][0];
            var currentEnd = intervals[i][1];

            if (currentStart < prevEnd) { // Overlap detected
                removedCount++; // This interval must be removed
            } else { // No overlap
                prevEnd = currentEnd; // Update prevEnd to current interval's end
            }
        }
        return removedCount;
    }
    // FAANG Tip: The greedy strategy of sorting by end time is optimal. It maximizes the remaining available space for subsequent intervals.

    public static void main(String[] args) {
        var solution = new NonOverlappingIntervals();

        // Test Case 1: Example from LeetCode
        var intervals1 = new int[][]{{1, 2}, {2, 3}, {3, 4}, {1, 3}};
        // Expected: 1 (Remove [1,3])
        System.out.println("Intervals: " + Arrays.deepToString(intervals1) + " -> Removed: " + solution.eraseOverlapIntervals(intervals1));

        // Test Case 2: Example from LeetCode
        var intervals2 = new int[][]{{1, 2}, {1, 2}, {1, 2}};
        // Expected: 2 (Keep one [1,2], remove two [1,2])
        System.out.println("Intervals: " + Arrays.deepToString(intervals2) + " -> Removed: " + solution.eraseOverlapIntervals(intervals2));

        // Test Case 3: Example from LeetCode
        var intervals3 = new int[][]{{1, 2}, {2, 3}};
        // Expected: 0 (No overlaps)
        System.out.println("Intervals: " + Arrays.deepToString(intervals3) + " -> Removed: " + solution.eraseOverlapIntervals(intervals3));

        // Test Case 4: Complex overlaps
        var intervals4 = new int[][]{{1, 100}, {11, 22}, {1, 11}, {2, 12}};
        // Expected: 2 (Keep [1,100], remove [1,11], [2,12] OR keep [11,22], remove [1,100], [1,11], [2,12] -> 3 removed.
        // Sorted by end: [1,11], [2,12], [11,22], [1,100]
        // Keep [1,11], prevEnd=11. [2,12] overlaps, remove. [11,22] no overlap, prevEnd=22. [1,100] overlaps, remove. Total 2 removed.
        System.out.println("Intervals: " + Arrays.deepToString(intervals4) + " -> Removed: " + solution.eraseOverlapIntervals(intervals4));

        // Test Case 5: Empty intervals
        var intervals5 = new int[][]{};
        // Expected: 0
        System.out.println("Intervals: " + Arrays.deepToString(intervals5) + " -> Removed: " + solution.eraseOverlapIntervals(intervals5));

        // Test Case 6: Single interval
        var intervals6 = new int[][]{{1, 5}};
        // Expected: 0
        System.out.println("Intervals: " + Arrays.deepToString(intervals6) + " -> Removed: " + solution.eraseOverlapIntervals(intervals6));
    }
}
